driver-legacy/src/main/com/mongodb/Java8DBCursorCleaner.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+
+/**
+ * A {@link Object#finalize()}-based implementation of {@link DBCursorCleaner}.
+ * {@link DBCursorCleaner#create(MongoClient, MongoNamespace, ServerCursor)} is responsible for ensuring
+ * that this class is only used if the {@code java.lang.ref.Cleaner} class is not available
+ * (i.e. the runtime is Java 8).
+ */
+final class Java8DBCursorCleaner extends DBCursorCleaner {
+    private final MongoClient mongoClient;
+    private final MongoNamespace namespace;
+    private ServerCursor serverCursor;
```

Comment:
I think that both here and in `CleanerState`, the field `serverCursor` must be `volatile`. The finalization mechanism / `Cleaner` does not guarantee that writes to the object that is being finalized / the object representing cleaning state & action happen before the execution of `finalize()` / cleaning action. At first I was afraid that our current implementation has the same problem, but it does not: `OptionalFinalizer.serverCursor` is `volatile`.

Reply:
I think you're right.  That was careless of me.

Done.